dali/kernels/imgproc/convolution/filter_gpu.cuh

```diff
@@ -0,0 +1,1231 @@
+          ivec2 offset{s, r + lanes_offset};
+          process_position(filter_coef, offset, lane);
+        }
+      }
+    }
+  }
+
+  /**
+   * @brief Iterates over all filter positions and produces offsets of the corresponding
+   * positions in the input. For each filter position, `lanes` offsets are computed
+   * to amortize the overhead of the for loops.
+   */
+  template <typename ProcessInputPosition>
+  DALI_DEVICE DALI_FORCEINLINE void compute_conv_lanes(
+      ivec3 filter_extents, int in_filter_width, ProcessInputPosition&& process_position) const {
+    ivec3 block_dim = block_setup_.block_dim();
+    const auto* filter = sample_desc_.filter;
+    for (int p = 0; p < filter_extents.z; p++) {
+      for (int r = 0; r < filter_extents.y; r++) {
+        for (int s = 0; s < in_filter_width; s += sample_desc_.in_shape.num_channels) {
+          auto filter_coef = __ldg(filter++);
+          ivec3 filter_position{s, r, p};
+#pragma unroll
+          for (int lane = 0; lane < StaticConfigT::lanes; lane++) {
+            process_position(filter_coef, filter_position, lane);
+            filter_position[lane_axis] += block_dim[lane_axis];
+          }
+        }
+      }
+    }
+  }
+
+  DALI_DEVICE DALI_FORCEINLINE void precompute_indices(ivec2 anchored_start) const {
+    for (int y = block_setup_.flat_idx(); y < sample_desc_.workspace_desc.in_extents.y;
+         y += block_setup_.flat_size()) {
+      precomputed_idx_[y] = in_loader_.border_remap(anchored_start.y + y, sample_desc_.in_shape, 1);
+    }
+  }
+
+  DALI_DEVICE DALI_FORCEINLINE void precompute_indices(ivec3 anchored_start) const {
+    int* ys = precomputed_idx_;
+    for (int y = block_setup_.flat_idx(); y < sample_desc_.workspace_desc.in_extents.y;
+         y += block_setup_.flat_size()) {
+      ys[y] = in_loader_.border_remap(anchored_start.y + y, sample_desc_.in_shape, 1);
+    }
+    int* zs = precomputed_idx_ + sample_desc_.workspace_desc.in_extents.y;
+    for (int z = block_setup_.flat_idx(); z < sample_desc_.workspace_desc.in_extents.z;
+         z += block_setup_.flat_size()) {
+      zs[z] = in_loader_.border_remap(anchored_start.z + z, sample_desc_.in_shape, 2);
+    }
+  }
+
+  DALI_DEVICE DALI_FORCEINLINE void load_input_to_shm(const In* __restrict__ in,
+                                                      ivec2 anchored_start) const {
+    ivec2 block_dim = block_setup_.block_dim();
+    ivec2 thread_idx = block_setup_.thread_idx();
+    for (int x = thread_idx.x; x < sample_desc_.workspace_desc.in_extents.x; x += block_dim.x) {
+      int global_x = in_loader_.border_remap_innermost(anchored_start.x + x, sample_desc_.in_shape);
+      for (int y = thread_idx.y; y < sample_desc_.workspace_desc.in_extents.y; y += block_dim.y) {
+        int global_y = precomputed_idx_[y];
+        in_workspace_[dot(ivec2{x, y}, sample_desc_.workspace_desc.in_strides)] =
+            in_loader_.load(in, ivec2{global_x, global_y}, sample_desc_.in_shape);
```

Comment:
The fact that `border_remap` is separated from `load`, but `load` still does some border handling (padding) seems a bit off...

Reply:
For smaller kernels, it seems that the way border handling is done makes a visible difference in the perf. Even without the precompute_idx thingy it is faster when decoupled from loading than when it is done in the load method in the innermost loop (sadly, as I hoped for compiler to hoist what can be hoisted). And to actually handle each extent once (with the precompute helper), I don't see other way.